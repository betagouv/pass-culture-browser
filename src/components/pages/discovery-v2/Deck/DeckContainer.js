import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import withSizes from 'react-sizes'
import { compose } from 'redux'

import Deck from './Deck'
import selectCurrentRecommendation from '../selectors/selectCurrentRecommendation'
import selectNextRecommendation from '../selectors/selectNextRecommendation'
import selectPreviousRecommendation from '../selectors/selectPreviousRecommendation'
import selectRecommendationsWithLastFakeReco from '../selectors/selectRecommendationsWithLastFakeRecommendation'
import { selectMediationById } from '../../../../redux/selectors/data/mediationsSelectors'
import { getNextLimit } from './utils/limits'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const { params } = match
  const { mediationId, offerId } = params

  const currentRecommendation = selectCurrentRecommendation(state, offerId, mediationId)
  const { mediationId: currentMediationId } = currentRecommendation || {}
  const currentMediation = selectMediationById(state, currentMediationId)
  const recommendations = selectRecommendationsWithLastFakeReco(state)
  const nextRecommendation = selectNextRecommendation(state, offerId, mediationId)
  const previousRecommendation = selectPreviousRecommendation(state, offerId, mediationId)

  const { thumbCount, tutoIndex } = currentMediation || {}
  const nbRecommendations = recommendations ? recommendations.length : 0
  const isTutoWithOnlyOneThumb = typeof tutoIndex === 'number' && thumbCount <= 1
  const hasNoVerso =
    !currentRecommendation || isTutoWithOnlyOneThumb || currentMediationId === 'fin'
  const nextLimit = getNextLimit(nbRecommendations)

  return {
    currentRecommendation,
    isFlipDisabled: hasNoVerso,
    nextLimit,
    nextRecommendation,
    previousRecommendation,
    recommendations,
  }
}

export const mapSizeToProps = ({ width, height }) => ({
  height,
  width: Math.min(width, 500),
})

export default compose(
  withRouter,
  withSizes(mapSizeToProps),
  connect(mapStateToProps)
)(Deck)
