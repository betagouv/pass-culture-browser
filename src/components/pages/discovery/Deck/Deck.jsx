import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import Draggable from 'react-draggable'

import getIsTransitionDetailsUrl from '../../../../utils/getIsTransitionDetailsUrl'
import getUrlWithoutDetailsPart from '../../../../utils/getUrlWithoutDetailsPart'
import isDetailsView from '../../../../utils/isDetailsView'
import CloseLink from '../../../layout/Header/CloseLink/CloseLink'
import CardContainer from './Card/CardContainer'
import NavigationContainer from './Navigation/NavigationContainer'

class Deck extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      versoPositionBeforeScroll: undefined,
      refreshKey: 0,
    }
  }

  componentDidMount() {
    const { currentRecommendation, recommendations } = this.props
    const isStateWithoutRecommendationsOrCurrentRecommendation =
      !recommendations || recommendations.length === 0 || !currentRecommendation

    if (!isStateWithoutRecommendationsOrCurrentRecommendation) {
      return
    }
    this.handleRefreshedDraggableKey()
  }

  componentDidUpdate() {
    const { history, location, match } = this.props
    const isTransitionDetailsUrl = getIsTransitionDetailsUrl(match)

    if (isTransitionDetailsUrl) {
      const { pathname, search } = location
      const nextUrl = `${pathname.split('/transition')[0]}${search}`
      history.replace(nextUrl)
    }
  }

  componentWillUnmount() {
    const { readTimeout, noDataTimeout } = this.props

    if (readTimeout) clearTimeout(readTimeout)
    if (noDataTimeout) clearTimeout(noDataTimeout)
  }

  handleOnStart = () => {
    const versoPositionBeforeScroll = document
      .getElementsByClassName('verso-header')[0]
      .getBoundingClientRect()
    this.setState({ versoPositionBeforeScroll })
  }

  handleOnStop = (event, data) => {
    const {
      currentRecommendation,
      height,
      horizontalSlideRatio,
      verticalSlideRatio,
      width,
    } = this.props
    const { versoPositionBeforeScroll } = this.state

    const index = (currentRecommendation && currentRecommendation.index) || 0
    const offset = (data.x + width * index) / width

    const isSwipingDownWhileAlreadyOnTop = versoPositionBeforeScroll.top >= -30 && data.y > 0
    if (isSwipingDownWhileAlreadyOnTop) {
      this.onHandleCloseCardDetails()
    } else if (data.y < -height * verticalSlideRatio) {
      this.handleShowCardDetails()
    } else if (offset > horizontalSlideRatio) {
      this.handleGoPrevious()
    } else if (-offset > horizontalSlideRatio) {
      this.handleGoNext()
    }
  }

  handleGoNext = () => {
    const { history, match, nextRecommendation } = this.props
    if (!nextRecommendation || isDetailsView(match)) {
      return
    }

    const { offerId, mediationId } = nextRecommendation
    const nextUrl = `/decouverte/${offerId}/${mediationId}`
    history.push(nextUrl)
    this.handleRefreshNext()
  }

  handleGoPrevious = () => {
    const { history, match, previousRecommendation } = this.props
    if (!previousRecommendation || isDetailsView(match)) {
      return
    }

    const { offerId, mediationId } = previousRecommendation
    const previousUrl = `/decouverte/${offerId}/${mediationId}`
    history.push(previousUrl)
  }

  handleRefreshNext = () => {
    const { currentRecommendation, handleRequestPutRecommendations, nextLimit } = this.props
    const hasReachedLimitBeforeFetchingRecommendations =
      nextLimit && currentRecommendation.index === nextLimit

    if (hasReachedLimitBeforeFetchingRecommendations) {
      handleRequestPutRecommendations()
    }
  }

  handleRefreshedDraggableKey = () => {
    this.setState(previousState => ({
      refreshKey: previousState.refreshKey + 1,
    }))
  }

  handleShowCardDetails = () => {
    const { history, location, match } = this.props
    const { pathname, search } = location
    if (isDetailsView(match)) {
      return
    }
    history.push(`${pathname}/details${search}`)
  }

  onHandleCloseCardDetails = () => {
    const { history, location, match } = this.props
    const removedDetailsUrl = getUrlWithoutDetailsPart(location, match)

    if (removedDetailsUrl) {
      history.push(removedDetailsUrl)
    }
    this.setState({ versoPositionBeforeScroll: undefined })
  }

  buildCloseToUrl = () => {
    const { location, match } = this.props
    return getUrlWithoutDetailsPart(location, match) + '/transition'
  }

  renderDraggableCards() {
    const {
      currentRecommendation,
      match,
      nextRecommendation,
      previousRecommendation,
      width,
    } = this.props
    const { index } = currentRecommendation || {}
    const { refreshKey } = this.state
    const position = {
      x: -1 * width * index,
      y: 0,
    }
    const draggableBounds = (isDetailsView(match) && {}) || {
      bottom: 0,
      left: position.x - width,
      right: position.x + width,
      top: -100,
    }

    return (
      <Draggable
        axis={isDetailsView(match) ? 'none' : 'both'}
        bounds={draggableBounds}
        enableUserSelectHack={false}
        key={refreshKey}
        onStart={this.handleOnStart}
        onStop={this.handleOnStop}
        position={position}
      >
        <div className="is-overlay">
          <div className="inner is-relative">
            {previousRecommendation && <CardContainer position="previous" />}
            <CardContainer position="current" />
            {nextRecommendation && <CardContainer position="next" />}
          </div>
        </div>
      </Draggable>
    )
  }

  scrollToTheTop = () => {
    document.getElementsByClassName('verso-wrapper')[0].scrollTop = 0
  }

  render() {
    const {
      currentRecommendation,
      height,
      match,
      nextRecommendation,
      previousRecommendation,
      recommendations,
    } = this.props
    const detailView = isDetailsView(match)
    const showNavigation = !detailView
    const nbRecommendations = recommendations.length

    return (
      <div
        className="is-clipped is-relative"
        data-nb-recos={nbRecommendations}
        id="deck"
      >
        {detailView && (
          <CloseLink
            actionOnClick={this.scrollToTheTop}
            closeTitle="Fermer"
            closeTo={this.buildCloseToUrl()}
          />
        )}

        {this.renderDraggableCards()}

        {showNavigation && currentRecommendation && (
          <NavigationContainer
            flipHandler={this.handleShowCardDetails || null}
            handleGoNext={(nextRecommendation && this.handleGoNext) || null}
            handleGoPrevious={(previousRecommendation && this.handleGoPrevious) || null}
            height={height}
            offerId={currentRecommendation.offerId}
          />
        )}
      </div>
    )
  }
}

Deck.defaultProps = {
  currentRecommendation: null,
  horizontalSlideRatio: 0.2,
  nextRecommendation: null,
  noDataTimeout: 20000,
  previousRecommendation: null,
  readTimeout: 2000,
  verticalSlideRatio: 0.1,
}

Deck.propTypes = {
  currentRecommendation: PropTypes.shape(),
  handleRequestPutRecommendations: PropTypes.func.isRequired,
  height: PropTypes.number.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
    replace: PropTypes.func.isRequired,
  }).isRequired,
  horizontalSlideRatio: PropTypes.number,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      details: PropTypes.string,
    }).isRequired,
  }).isRequired,
  nextLimit: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]).isRequired,
  nextRecommendation: PropTypes.shape(),
  noDataTimeout: PropTypes.number,
  previousRecommendation: PropTypes.shape(),
  readTimeout: PropTypes.number,
  recommendations: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  verticalSlideRatio: PropTypes.number,
  width: PropTypes.number.isRequired,
}

export default Deck
