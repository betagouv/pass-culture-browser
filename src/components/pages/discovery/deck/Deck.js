import get from 'lodash.get'
import { Icon } from 'pass-culture-shared'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Draggable from 'react-draggable'

import Card from '../card'
import DeckNavigation from '../DeckNavigation'
import { shouldShowVerso } from '../../../../helpers'
import {
  closeCardDetails,
  flipUnflippable,
  showCardDetails,
} from '../../../../reducers/card'

class Deck extends Component {
  constructor(props) {
    super(props)
    this.currentReadRecommendationId = null
    this.state = { refreshKey: 0 }
  }

  componentDidMount() {
    const { currentRecommendation, recommendations } = this.props
    this.handleUrlFlip()

    const isStateWithoutRecommendationsOrCurrentRecommendation =
      !recommendations || recommendations.length === 0 || !currentRecommendation

    if (!isStateWithoutRecommendationsOrCurrentRecommendation) return
    this.handleRefreshedDraggableKey()
  }

  componentDidUpdate() {
    const { match } = this.props
    const isVersoView = shouldShowVerso(match)
    // prevents to be called twice at Mount and at Update
    if (isVersoView) return
    this.handleUrlFlip()
  }

  componentWillUnmount() {
    const { dispatch, readTimeout, noDataTimeout } = this.props
    dispatch(closeCardDetails())

    if (readTimeout) clearTimeout(readTimeout)
    if (noDataTimeout) clearTimeout(noDataTimeout)
  }

  onStop = (e, data) => {
    const {
      width,
      height,
      draggable,
      verticalSlideRatio,
      horizontalSlideRatio,
      currentRecommendation,
    } = this.props

    const index = get(currentRecommendation, 'index', 0)
    const offset = (data.x + width * index) / width
    if (draggable && data.y > height * verticalSlideRatio) {
      this.handleCloseCardDetails()
    } else if (data.y < -height * verticalSlideRatio) {
      this.handleShowCardDetails()
    } else if (offset > horizontalSlideRatio) {
      this.handleGoPrevious()
    } else if (-offset > horizontalSlideRatio) {
      this.handleGoNext()
    }
  }

  handleGoNext = () => {
    const { history, areDetailsVisible, nextRecommendation } = this.props
    if (!nextRecommendation || areDetailsVisible) return
    const { offerId, mediationId } = nextRecommendation
    history.push(
      `/decouverte/${offerId || 'tuto'}${(mediationId && `/${mediationId}`) ||
        ''}`
    )
    this.handleRefreshNext()
  }

  handleGoPrevious = () => {
    const { history, areDetailsVisible, previousRecommendation } = this.props
    if (!previousRecommendation || areDetailsVisible) return
    const { offerId, mediationId } = previousRecommendation
    history.push(
      `/decouverte/${offerId || 'tuto'}${(mediationId && `/${mediationId}`) ||
        ''}`
    )
  }

  handleRefreshNext = () => {
    const { currentRecommendation, handleDataRequest, nextLimit } = this.props
    const shouldRequest = nextLimit && currentRecommendation.index === nextLimit
    if (!shouldRequest) return
    handleDataRequest()
  }

  handleRefreshedDraggableKey = () => {
    this.setState(previousState => ({
      refreshKey: previousState.refreshKey + 1,
    }))
  }

  handleShowCardDetails = () => {
    const { dispatch, isFlipDisabled } = this.props
    if (isFlipDisabled) return
    dispatch(showCardDetails())
  }

  handleCloseCardDetails = () => {
    const { dispatch, unFlippable } = this.props
    if (unFlippable) return
    dispatch(closeCardDetails())
  }

  handleUrlFlip = () => {
    const { dispatch, match } = this.props
    const isVersoView = shouldShowVerso(match)
    if (!isVersoView) return
    dispatch(flipUnflippable())
  }

  renderDraggableCards() {
    const {
      areDetailsVisible,
      currentRecommendation,
      nextRecommendation,
      previousRecommendation,
      width,
    } = this.props
    const { index } = currentRecommendation || {}
    const { refreshKey } = this.state

    const position = {
      x: -1 * width * index,
      y: 0,
    }
    const draggableBounds = (areDetailsVisible && {}) || {
      bottom: 0,
      left: position.x - width,
      right: position.x + width,
      top: -100,
    }

    return (
      <Draggable
        speed={{ x: 5 }}
        key={refreshKey}
        position={position}
        onStop={this.onStop}
        bounds={draggableBounds}
        enableUserSelectHack={false}
        axis={areDetailsVisible ? 'none' : 'exclude'}
      >
        <div className="is-overlay">
          <div className="inner is-relative">
            {previousRecommendation && <Card position="previous" />}
            <Card position="current" />
            {nextRecommendation && <Card position="next" />}
          </div>
        </div>
      </Draggable>
    )
  }

  render() {
    const {
      currentRecommendation,
      height,
      nextRecommendation,
      isFlipDisabled,
      areDetailsVisible,
      previousRecommendation,
      recommendations,
      unFlippable,
    } = this.props

    const showCloseButton = areDetailsVisible && !unFlippable
    const showNavigation = !areDetailsVisible || isFlipDisabled

    return (
      <div
        id="deck"
        className="is-clipped is-relative"
        data-nb-recos={recommendations.length}
      >
        {showCloseButton && (
          <button
            type="button"
            className="close-button"
            id="deck-close-verso-button"
            onClick={this.handleCloseCardDetails}
            style={{ zIndex: 300 }}
          >
            <Icon svg="ico-close" alt="Fermer" />
          </button>
        )}
        {this.renderDraggableCards()}
        {showNavigation && currentRecommendation && (
          <DeckNavigation
            recommendation={currentRecommendation}
            flipHandler={
              (!isFlipDisabled && this.handleShowCardDetails) || null
            }
            handleGoNext={(nextRecommendation && this.handleGoNext) || null}
            handleGoPrevious={
              (previousRecommendation && this.handleGoPrevious) || null
            }
            height={height}
          />
        )}
      </div>
    )
  }
}

Deck.defaultProps = {
  currentRecommendation: null,
  horizontalSlideRatio: 0.2,
  nextRecommendation: null,
  noDataTimeout: 20000,
  previousRecommendation: null,
  readTimeout: 2000,
  verticalSlideRatio: 0.1,
}

Deck.propTypes = {
  areDetailsVisible: PropTypes.bool.isRequired,
  currentRecommendation: PropTypes.object,
  dispatch: PropTypes.func.isRequired,
  draggable: PropTypes.bool.isRequired,
  handleDataRequest: PropTypes.func.isRequired,
  height: PropTypes.number.isRequired,
  history: PropTypes.object.isRequired,
  horizontalSlideRatio: PropTypes.number,
  isFlipDisabled: PropTypes.bool.isRequired,
  match: PropTypes.object.isRequired,
  nextLimit: PropTypes.oneOfType([PropTypes.bool, PropTypes.number]).isRequired,
  nextRecommendation: PropTypes.object,
  noDataTimeout: PropTypes.number,
  previousRecommendation: PropTypes.object,
  readTimeout: PropTypes.number,
  recommendations: PropTypes.array.isRequired,
  unFlippable: PropTypes.bool.isRequired,
  verticalSlideRatio: PropTypes.number,
  width: PropTypes.number.isRequired,
}

export default Deck
