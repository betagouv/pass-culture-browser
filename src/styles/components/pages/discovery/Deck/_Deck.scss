#deck {
  height: 100%;
  max-height: 100%;
  background-color: black;

  .react-draggable {
    z-index: $zindex-deckdraggable;
    cursor: -webkit-grab;
    transition-property: transform;
    transition-timing-function: ease;
    transition-duration: 250ms;
  }

  .react-draggable .inner {
    width: 100%;
    height: 100%;
  }

  .close {
    top: 1rem;
    right: 1rem;
  }
}

.android-with-statusbar .close {
  top: 2rem;
}
