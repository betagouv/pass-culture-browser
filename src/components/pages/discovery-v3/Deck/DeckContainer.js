import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import withSizes from 'react-sizes'
import { compose } from 'redux'

import Deck from './Deck'
import selectCurrentRecommendation from '../selectors/selectCurrentRecommendation'
import selectNextRecommendation from '../selectors/selectNextRecommendation'
import selectPreviousRecommendation from '../selectors/selectPreviousRecommendation'
import selectUniqAndIndexifiedRecommendations from '../selectors/selectUniqAndIndexifiedRecommendations'
import { getNextLimit } from './utils/limits'

export const mapStateToProps = (state, ownProps) => {
  const { match } = ownProps
  const { params } = match
  const { mediationId, offerId } = params

  const currentRecommendation = selectCurrentRecommendation(state, offerId, mediationId)
  const recommendations = selectUniqAndIndexifiedRecommendations(state)
  const nextRecommendation = selectNextRecommendation(state, offerId, mediationId)
  const previousRecommendation = selectPreviousRecommendation(state, offerId, mediationId)

  const nbRecommendations = recommendations ? recommendations.length : 0
  const hasNoVerso = !currentRecommendation
  const nextLimit = getNextLimit(nbRecommendations)

  return {
    currentRecommendation,
    isFlipDisabled: hasNoVerso,
    nextLimit,
    nextRecommendation,
    previousRecommendation,
    recommendations,
  }
}

export const mapSizeToProps = ({ width, height }) => ({
  height,
  width: Math.min(width, 500),
})

export default compose(
  withRouter,
  withSizes(mapSizeToProps),
  connect(mapStateToProps)
)(Deck)
