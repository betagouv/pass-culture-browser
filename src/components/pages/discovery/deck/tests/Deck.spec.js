import React from 'react'
import { shallow } from 'enzyme'

import Deck from '../Deck'

const dispatchMock = jest.fn()
const handleDataRequestMock = jest.fn()

describe('src | components | pages | discovery | deck |Â Deck', () => {
  const initialProps = {
    areDetailsVisible: false,
    backButton: true,
    currentRecommendation: {},
    dispatch: dispatchMock,
    draggable: true,
    handleDataRequest: handleDataRequestMock,
    height: 947,
    history: {
      location: {
        search: '',
      },
    },
    isFlipDisabled: false,
    match: { params: {} },
    nextLimit: 50,
    previousLimit: 40,
    recommendations: [{}],
    unFlippable: false,
    width: 500,
  }

  describe('snapshot', () => {
    it('should match snapshot', () => {
      // when
      const wrapper = shallow(<Deck {...initialProps} />)

      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('react functions', () => {
    describe('constructor', () => {
      it('should initialize state correctly', () => {
        // when
        const wrapper = shallow(<Deck {...initialProps} />)
        const expected = {
          refreshKey: 0,
        }

        // then
        expect(wrapper.state()).toEqual(expected)
        expect(wrapper.instance().currentReadRecommendationId).toEqual(null)
      })
    })

    describe('componentDidMount', () => {
      describe('Whenever', () => {
        it('should call handleUrlFlip', async () => {
          // given
          const props = {
            areDetailsVisible: false,
            backButton: true,
            dispatch: dispatchMock,
            draggable: true,
            handleDataRequest: handleDataRequestMock,
            height: 947,
            history: {
              location: {
                search: '',
              },
            },
            isFlipDisabled: false,
            match: {
              params: {
                mediationId: 'HM',
                offerId: 'KQ',
              },
            },
            nextLimit: 50,
            previousLimit: 40,
            recommendations: [],
            unFlippable: false,
            width: 500,
          }

          // when
          const wrapper = shallow(<Deck {...props} />)
          const wrapperInstance = wrapper.instance()
          const spy = jest
            .spyOn(wrapperInstance, 'handleUrlFlip')
            .mockImplementation(() => {})
          wrapper.setProps(props)
          // setProps() make the componentDidMount after instanciating the spy

          // then
          expect(spy).toHaveBeenCalled()
        })
      })

      describe('When there is recommendations', () => {
        it('should not refresh the key of draggable component', () => {
          // given
          const props = {
            areDetailsVisible: false,
            backButton: true,
            currentRecommendation: {
              bookingsIds: [],
            },
            dispatch: dispatchMock,
            draggable: true,
            handleDataRequest: handleDataRequestMock,
            height: 947,
            history: {
              location: {
                search: '',
              },
            },
            isFlipDisabled: false,
            match: {
              params: {
                mediationId: 'HM',
                offerId: 'KQ',
              },
            },
            nextLimit: 50,
            previousLimit: 40,
            recommendations: [{}],
            unFlippable: false,
            width: 500,
          }

          // when
          const wrapper = shallow(<Deck {...props} />)

          // then
          expect(wrapper.state()).toEqual({ refreshKey: 0 })
        })
      })

      describe('When there is no recommendations or currentRecommendation available', () => {
        it('should call handleRefreshedDraggableKey', () => {
          // given
          const props = {
            areDetailsVisible: false,
            backButton: true,
            dispatch: dispatchMock,
            draggable: true,
            handleDataRequest: handleDataRequestMock,
            height: 947,
            history: {
              location: {
                search: '',
              },
            },
            isFlipDisabled: false,
            match: {
              params: {
                mediationId: 'HM',
                offerId: 'KQ',
              },
            },
            nextLimit: 50,
            previousLimit: 40,
            recommendations: [],
            unFlippable: false,
            width: 500,
          }

          // when
          const wrapper = shallow(<Deck {...props} />)
          wrapper.setProps(props)

          // then
          expect(wrapper.state()).toEqual({ refreshKey: 1 })
        })
      })
    })

    describe('componentWillUnmount', () => {
      it('should dispatch unFlip', () => {
        // when
        const wrapper = shallow(<Deck {...initialProps} />)
        wrapper.unmount()

        // then
        expect(dispatchMock.mock.calls.length).toBe(1)
        expect(dispatchMock).toHaveBeenCalledWith({
          type: 'CLOSE_DETAILS_VIEW',
        })
      })

      it('should clearTimeout', () => {
        jest.useFakeTimers()
        // when
        const wrapper = shallow(<Deck {...initialProps} />)
        wrapper.unmount()

        // then
        expect(clearTimeout).toHaveBeenCalled()
      })
    })
  })
})
