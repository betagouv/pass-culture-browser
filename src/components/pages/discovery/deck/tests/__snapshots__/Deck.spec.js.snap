// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | discovery | deck |Â Deck snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Deck
    areDetailsVisible={false}
    backButton={true}
    currentRecommendation={Object {}}
    dispatch={[MockFunction]}
    draggable={true}
    handleDataRequest={[MockFunction]}
    height={947}
    history={
      Object {
        "location": Object {
          "search": "",
        },
      }
    }
    horizontalSlideRatio={0.2}
    isFlipDisabled={false}
    match={
      Object {
        "params": Object {},
      }
    }
    nextLimit={50}
    nextRecommendation={null}
    noDataTimeout={20000}
    previousLimit={40}
    previousRecommendation={null}
    readTimeout={2000}
    recommendations={
      Array [
        Object {},
      ]
    }
    unFlippable={false}
    verticalSlideRatio={0.1}
    width={500}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <Draggable
          allowAnyClick={false}
          axis="exclude"
          bounds={
            Object {
              "bottom": 0,
              "left": NaN,
              "right": NaN,
              "top": -100,
            }
          }
          cancel={null}
          defaultClassName="react-draggable"
          defaultClassNameDragged="react-draggable-dragged"
          defaultClassNameDragging="react-draggable-dragging"
          defaultPosition={
            Object {
              "x": 0,
              "y": 0,
            }
          }
          disabled={false}
          enableUserSelectHack={false}
          grid={null}
          handle={null}
          offsetParent={null}
          onDrag={[Function]}
          onMouseDown={[Function]}
          onStart={[Function]}
          onStop={[Function]}
          position={
            Object {
              "x": NaN,
              "y": 0,
            }
          }
          speed={
            Object {
              "x": 5,
            }
          }
          transform={null}
        >
          <div
            className="is-overlay"
          >
            <div
              className="inner is-relative"
            >
              <withSizes(withRouter(Connect(RawCard)))
                position="current"
              />
            </div>
          </div>
        </Draggable>,
        <withRouter(Connect(RawDeckNavigation))
          flipHandler={[Function]}
          handleGoNext={null}
          handleGoPrevious={null}
          height={947}
          recommendation={Object {}}
        />,
      ],
      "className": "is-clipped is-relative",
      "data-nb-recos": 1,
      "id": "deck",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "allowAnyClick": false,
          "axis": "exclude",
          "bounds": Object {
            "bottom": 0,
            "left": NaN,
            "right": NaN,
            "top": -100,
          },
          "cancel": null,
          "children": <div
            className="is-overlay"
          >
            <div
              className="inner is-relative"
            >
              <withSizes(withRouter(Connect(RawCard)))
                position="current"
              />
            </div>
          </div>,
          "defaultClassName": "react-draggable",
          "defaultClassNameDragged": "react-draggable-dragged",
          "defaultClassNameDragging": "react-draggable-dragging",
          "defaultPosition": Object {
            "x": 0,
            "y": 0,
          },
          "disabled": false,
          "enableUserSelectHack": false,
          "grid": null,
          "handle": null,
          "offsetParent": null,
          "onDrag": [Function],
          "onMouseDown": [Function],
          "onStart": [Function],
          "onStop": [Function],
          "position": Object {
            "x": NaN,
            "y": 0,
          },
          "speed": Object {
            "x": 5,
          },
          "transform": null,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="inner is-relative"
            >
              <withSizes(withRouter(Connect(RawCard)))
                position="current"
              />
            </div>,
            "className": "is-overlay",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                null,
                <withSizes(withRouter(Connect(RawCard)))
                  position="current"
                />,
                null,
              ],
              "className": "inner is-relative",
            },
            "ref": null,
            "rendered": Array [
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "position": "current",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              null,
            ],
            "type": "div",
          },
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "flipHandler": [Function],
          "handleGoNext": null,
          "handleGoPrevious": null,
          "height": 947,
          "recommendation": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <Draggable
            allowAnyClick={false}
            axis="exclude"
            bounds={
              Object {
                "bottom": 0,
                "left": NaN,
                "right": NaN,
                "top": -100,
              }
            }
            cancel={null}
            defaultClassName="react-draggable"
            defaultClassNameDragged="react-draggable-dragged"
            defaultClassNameDragging="react-draggable-dragging"
            defaultPosition={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            disabled={false}
            enableUserSelectHack={false}
            grid={null}
            handle={null}
            offsetParent={null}
            onDrag={[Function]}
            onMouseDown={[Function]}
            onStart={[Function]}
            onStop={[Function]}
            position={
              Object {
                "x": NaN,
                "y": 0,
              }
            }
            speed={
              Object {
                "x": 5,
              }
            }
            transform={null}
          >
            <div
              className="is-overlay"
            >
              <div
                className="inner is-relative"
              >
                <withSizes(withRouter(Connect(RawCard)))
                  position="current"
                />
              </div>
            </div>
          </Draggable>,
          <withRouter(Connect(RawDeckNavigation))
            flipHandler={[Function]}
            handleGoNext={null}
            handleGoPrevious={null}
            height={947}
            recommendation={Object {}}
          />,
        ],
        "className": "is-clipped is-relative",
        "data-nb-recos": 1,
        "id": "deck",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "allowAnyClick": false,
            "axis": "exclude",
            "bounds": Object {
              "bottom": 0,
              "left": NaN,
              "right": NaN,
              "top": -100,
            },
            "cancel": null,
            "children": <div
              className="is-overlay"
            >
              <div
                className="inner is-relative"
              >
                <withSizes(withRouter(Connect(RawCard)))
                  position="current"
                />
              </div>
            </div>,
            "defaultClassName": "react-draggable",
            "defaultClassNameDragged": "react-draggable-dragged",
            "defaultClassNameDragging": "react-draggable-dragging",
            "defaultPosition": Object {
              "x": 0,
              "y": 0,
            },
            "disabled": false,
            "enableUserSelectHack": false,
            "grid": null,
            "handle": null,
            "offsetParent": null,
            "onDrag": [Function],
            "onMouseDown": [Function],
            "onStart": [Function],
            "onStop": [Function],
            "position": Object {
              "x": NaN,
              "y": 0,
            },
            "speed": Object {
              "x": 5,
            },
            "transform": null,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="inner is-relative"
              >
                <withSizes(withRouter(Connect(RawCard)))
                  position="current"
                />
              </div>,
              "className": "is-overlay",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                  <withSizes(withRouter(Connect(RawCard)))
                    position="current"
                  />,
                  null,
                ],
                "className": "inner is-relative",
              },
              "ref": null,
              "rendered": Array [
                null,
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "position": "current",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                null,
              ],
              "type": "div",
            },
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "flipHandler": [Function],
            "handleGoNext": null,
            "handleGoPrevious": null,
            "height": 947,
            "recommendation": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
